### ISSUE TITLE
How to customize which characters are considered higher in the alphabet than others?

### ISSUE THREAD

**Issue by emmahsax** [1nVitr0/plugin-vscode-blocksort#84]
I don't fully understand the collation rules, but all I know is the following blocks would be sorted like this:

```
variable "grafana_address" {
  type = string
}
variable "grafana_password" {
  type = string
}
variable "grafana-count" {
  type = number
}
variable "grafana-username" {
  type = string
}
variable "grafana" {
  type = bool
}
```

But I'd like them to be sorted like this:

```
variable "grafana" {
  type = bool
}
variable "grafana-count" {
  type = number
}
variable "grafana-username" {
  type = string
}
variable "grafana_address" {
  type = string
}
variable "grafana_password" {
  type = string
}
```

So essentially the `_` should be considered less far down the alphabet than the letters, and the `-` should be considered higher up the alphabet than `_`. Are these types of customizations possible?

**Comment by 1nVitr0** [comment]
Yes, that is possible through the exteded `collatorOptions` as described in the [Extension Settings](https://github.com/1nVitr0/plugin-vscode-blocksort#extension-settings) (though admittedly not very clearly):

> - `collatorOptions`: Collator options for sorting. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator/Collator for more information
>   - Default:
>     ```json
>     {
>       "numeric": true,
>       "caseFirst": "false",
>       "sensitivity": "base"
>     }
>     ```
>   - In addition to the default JavaScript Collator options, the following properties can be supplied
>     - `locales`: A BCP 47 language tag, or an comma separated array of such strings.
>     - `customSortOrder`: Custom Sort order in the form of a list of characters

This means, in order to sort `_` before `-`, it should be enough to add the following to your settings:

```json
{
    "numeric": true,
    "caseFirst": "false",
    "sensitivity": "base",
    "customSortOrder": "_-"
}
```

If that doesn't work, let me know :)

**Comment by emmahsax** [comment]
@1nVitr0 That still doesn't seem to work. I put those settings into my VS Code, and then uninstalled the extension and reinstalled it.

**Comment by emmahsax** [comment]
Thank you. This seems to work brilliantly.

The one thing I did have to add was the alphabet. So my final custom order is this (I also set `caseFirst` to be `upper`, but that was my decision outside of this):

```json
"blocksort.collatorOptions": {
  "numeric": true,
  "caseFirst": "upper",
  "sensitivity": "base",
  "customSortOrder": "abcdefghijklmnopqrstuvwxyz\"-_"
},
```

But all of this does exactly what I want it to. Thank you so much for the guidance and quick fix!

<details><summary><b>For more info as to why I added the alphabet...</b></summary>

For some reason, the `"` part isn't behaving how I'd expect.

Here's an example where it's not working:

```
# Actual
variable "grafana" {
  type = bool
}
variable "grafana_password" {
  type = string
}
variable "name" {
  type = string
}
variable "helm_chart_version" {
  type = string
}

# Expected
variable "grafana" {
  type = bool
}
variable "grafana_password" {
  type = string
}
variable "helm_chart_version" {
  type = string
}
variable "name" {
  type = string
}
```

And an example where it is working:

```
variable "grafana" {
  type = bool
}
variable "grafana-count" {
  type = number
}
variable "grafana-username" {
  type = string
}
variable "grafana_address" {
  type = string
}
variable "grafana_password" {
  type = string
}
variable "name" {
  type = string
}
```

I've narrowed down to something about the character count. Because `name` and `helm` are the same amount of characters, it's reading `"name"` to come before `"helm_release"`, even though `h` comes before `n`, simply because it's reading `"` to come before `_`. But if I add the alphabet at the beginning, then it properly recognizes that `h` should come before `n`.

</details>

**Comment by 1nVitr0** [comment]
That makes total sense, thanks for the detailed info! There is already an  inbuilt option for collators to ignore punctuation, I might just extend that to quotes and brackets by default. This is probably a frequent use case for custom sort order.
